### PR TITLE
Hq dropdowns

### PR DESCRIPTION
**Pull request by sravfeyn**
Implements https://docs.google.com/a/dimagi.com/document/d/1GtVw-_Wq7aRyFCY0ervt6Q8nQNsJde5R_zZXFXAfr-Q/edit?disco=AAAAAQ_ZnoE

More than half of the commits are about performance though. The main navigational-menu loading time increases by ~0.140 seconds from around 0.09 seconds to 0.23 seconds. 

@czue @snopoke - the number of calls to various db-level queries (Domain.get_by_name) is reduced to 2 from around 37.

### CODE REVIEW

**Comment by sravfeyn** [comment]
Hey @czue @snopoke This has passed QA and is good to go.

File: corehq/apps/hqwebapp/models.py [hunk 1]
```diff
@@ -1034,23 +1096,26 @@ def web_username(request=None, couch_user=None, **context):
...
-                     },
-                     {
-                         'title': _('My Information'),
-                         'urlname': EditMyAccountDomainView.urlname
-                     }
-                 ]}
+                {
+                    'title': get_web_user_list_view(self._request).page_title,
+                    'url': reverse(get_web_user_list_view(self._request).urlname, args=[self.domain]),
+                    'description': _("Grant other CommCare HQ users access to your project and manage user roles."),
```
Comment (dimagimon) [inline corehq/apps/hqwebapp/models.py]: E501 line too long (116 > 115 characters)

File: corehq/apps/hqwebapp/utils.py [hunk 2]
```diff
@@ -165,3 +165,166 @@ def get_bulk_upload_form(context, context_key="bulk_upload"):
...
+        if (subpage.get('show_in_dropdown', False) and
+           not subpage.get('show_in_first_level', False)):
+            second_level_dropdowns.append(format_submenu_context(
+                subpage['title'],
+                url=reverse(subpage['urlname'], args=[domain])),
+            )
+    return second_level_dropdowns
+
+
+def get_first_level_dropdowns(subpages, domain=None):
```
Comment (snopoke) [inline corehq/apps/hqwebapp/utils.py]: This seems identical to `get_second_level_dropdowns` except for variable names.

File: corehq/apps/hqwebapp/utils.py [hunk 3]
```diff
@@ -165,3 +165,166 @@ def get_bulk_upload_form(context, context_key="bulk_upload"):
...
+
+
+def get_second_level_dropdowns(subpages, domain=None):
+    """
+        formats subpages of a sidebar_item as second level dropdown items
+    """
+    second_level_dropdowns = []
+    for subpage in subpages:
+        if (subpage.get('show_in_dropdown', False) and
+           not subpage.get('show_in_first_level', False)):
```
Comment (snopoke) [inline corehq/apps/hqwebapp/utils.py]: Should this maybe say `show_in_second_level`?

File: corehq/apps/hqwebapp/utils.py (hunk 2 above)
Comment (snopoke) [inline corehq/apps/hqwebapp/utils.py]: Could refactor to use the same function

**Comment by snopoke** [comment]
Generally looks good, see specific comments. Lost of unrelated formatting changes, did you use an auto-formatter?

File: corehq/apps/hqwebapp/utils.py (hunk 3 above)
Comment (sravfeyn) [inline corehq/apps/hqwebapp/utils.py]: Each link in the sidebar (say Mobile Workers) has subpages (Add Mobile Worker for Mobile Workers). These won't be shown in Dropdown menu, but if `show_in_dropdown` is `True` then it's shown as a second-level-dropdown, unless `show_in_first_level` is explicitly set to `True`. So there is no setting called `show_in_second_level`.

**Comment by sravfeyn** [comment]
Regarding formatting changes, I am not using any auto-formatter, can you point me where you see those?

File: corehq/apps/hqwebapp/utils.py (hunk 2 above)
Comment (sravfeyn) [inline corehq/apps/hqwebapp/utils.py]: The difference is in 3rd line of for-loop, `not subpage.get('show_in_first_level', False)):`. Can refactor into one single function, but then we might need to pass an extra var like `level=1` or `level=2`. That didn't seem neater, should I refactor into one function?

Comment (sravfeyn) [inline corehq/apps/hqwebapp/utils.py]: Actually it is more neater, will refactor.

**Comment by dimagimon** [comment]
There are 4 errors:

* corehq/apps/hqwebapp/utils.py, line 269 - E225 missing whitespace around operator
* corehq/apps/hqwebapp/utils.py, line 271 - E225 missing whitespace around operator
* corehq/apps/hqwebapp/utils.py, line 275 - E126 continuation line over-indented for hanging indent
* corehq/apps/hqwebapp/utils.py, line 277 - E126 continuation line over-indented for hanging indent

**Comment by snopoke** [comment]
Looks good, waiting for build.

File: corehq/apps/hqwebapp/models.py [hunk 4]
```diff
@@ -214,13 +203,31 @@ def subpage_url_names(self):
         try:
             for name, section in self.sidebar_items:
                 names.extend(subpage['urlname']
-                    for item in section
-                    for subpage in item.get('subpages', []))
+                             for item in section
+                             for subpage in item.get('subpages', []))
```
Comment (snopoke) [inline corehq/apps/hqwebapp/models.py]: This is an example of a formatting change that is unnecessary. Even though this is an improvement it just makes it harder to see what are real changes and what is just formatting. Generally you should try and leave formatting as is unless you are already editing that piece of code.

If it is really bugging you you can always do a PR with just formatting changes.

**Comment by mwhite** [comment]
lol, @czue and @dannyroberts, remember when I did this 2 years ago at the same time as the initial dynamic navigation functionality, and @czue was like "B-b-but is that what you thought you were supposed to do?"

**Pull request merged**